ocaml/xapi/import_raw_vdi.ml

```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
               debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
               if (Importexport.check_sr_availability ~__context sr)
               then localhost_handler rpc session_id vdi req s
-              else
-                let host = Importexport.find_host_for_sr ~__context sr in
-                let address = Db.Host.get_address ~__context ~self:host in
-                return_302_redirect req s address;
-                None
+              else 
+                begin
+                  try 
+                    let host = Importexport.find_host_for_sr ~__context sr in
+                    let address = Db.Host.get_address ~__context ~self:host in
+                    return_302_redirect req s address;
+                    None
+                  with 
+                  | Api_errors.Server_error (a,b) as e ->
+                    error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+                    let all = req.Request.cookie @ req.Request.query in 
+                    let task_id = 
+                    if List.mem_assoc "task_id" all 
+                    then Some (Ref.of_string (List.assoc "task_id" all))
+                    else None in
+                    begin match task_id with
+                    | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
+                    | Some task_id -> Server_helpers.exec_with_forwarded_task task_id (fun __context -> TaskHelper.failed ~__context e)
+                    end;
+                    Http_svr.headers s (Http.http_400_badrequest ());
+                    None
+                  | e ->
+                    error "Caught exception in import handler: %s" (Printexc.to_string e);                   
+                    raise e
+                end;
             | None ->
               error "Require an SR or VDI to import";
-              TaskHelper.failed ~__context Api_errors.(Server_error(vdi_missing,[]));
+              let e = Api_errors.(Server_error(vdi_missing,[])) in 
+              let all = req.Request.cookie @ req.Request.query in
+              let task_id =
+              if List.mem_assoc "task_id" all 
+              then Some (Ref.of_string (List.assoc "task_id" all))
+              else None in
+              begin match task_id with
+              | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
```

Comment:
If I'm not mistaken, we do not need this line, because it will set a newly-created task to failed, which is not useful for the caller?
Oh, or I assume you created this task to indicate that something failed. In that case, I think we don't need to create a new task here, maybe we could just use the context from line 101: ` Server_helpers.exec_with_new_task "VDI.import" ...` and set that to failed.